pkg/reconciler/taskrun/resources/taskspec.go

```diff
@@ -85,10 +91,53 @@ func GetTaskData(ctx context.Context, taskRun *v1.TaskRun, getTask GetTask) (*re
 		return nil, nil, fmt.Errorf("taskRun %s not providing TaskRef or TaskSpec", taskRun.Name)
 	}
 
+	steps, err := extractStepActions(ctx, taskSpec, taskRun, getStepAction)
+	if err != nil {
+		return nil, nil, err
+	} else {
+		taskSpec.Steps = steps
+	}
+
 	taskSpec.SetDefaults(ctx)
 	return &resolutionutil.ResolvedObjectMeta{
 		ObjectMeta:         &taskMeta,
 		RefSource:          refSource,
 		VerificationResult: verificationResult,
 	}, &taskSpec, nil
 }
+
+// extractStepActions extracts the StepActions and merges them with the inlined Step specification.
+func extractStepActions(ctx context.Context, taskSpec v1.TaskSpec, taskRun *v1.TaskRun, getStepAction GetStepAction) ([]v1.Step, error) {
+	steps := []v1.Step{}
+	for _, step := range taskSpec.Steps {
+		s := step.DeepCopy()
+		if step.Ref != nil {
+			if !config.FromContextOrDefaults(ctx).FeatureFlags.EnableStepActions {
+				return nil, apis.ErrGeneric("feature flag %s should be set to true to reference StepActions in Steps.", config.EnableStepActions)
```

Comment:
I don't see any code in reconciler returning apis.ErrGeneric, it is mainly used in webhooks. 
And do we need to validate this again here considering that we already have it in validation webhooks? And that validation function will also be invoked when creating the pod?

Reply:
Yes, I was just a bit paranoid. I can remove that check completely from the controller side.

Reply:
fixed.